### PR TITLE
Slipper

### PR DESCRIPTION
**Pull request by sms1097**

### CODE REVIEW

**Comment by sms1097** [comment]
This is definitely an initial commit. Let me know if you see some opportunities for better design choices/catching edge cases.

File: imodels/rule_set/boosted_rules.py [hunk 1]
```diff
@@ -18,9 +19,13 @@ class BoostedRulesClassifier(RuleSet, BaseEstimator, MetaEstimatorMixin, Classif
     Currently limited to only binary classification.
     '''
 
-    def __init__(self, n_estimators=10, estimator=partial(DecisionTreeClassifier, max_depth=1)):
+    def __init__(self, method='DT', n_estimators=10):
```
Comment (sms1097) [inline imodels/rule_set/boosted_rules.py:22]: Not sure what a good design for handling depreciation of estimator as partial instead of string here. This approach will cause problems with existing code on user end.

File: imodels/rule_set/slipper.py [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+        else:
+            new_feats = features
+
+        self.feature_dict = {
+            old_feat:new_feat for old_feat, new_feat in enumerate(new_feats)
+        }
+
+    def predict_proba(self, X):
+
+        proba = self.predict(X)
```
Comment (sms1097) [inline imodels/rule_set/slipper.py:250]: Not entirely sure what predict proba should return in the case of slipper. This just returns array of shape (n,2) where column 1 is raw predictions.

File: imodels/rule_set/slipper.py [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+
+from joblib import Parallel, delayed
+
+from sklearn.base import BaseEstimator, ClassifierMixin
+from sklearn.model_selection import train_test_split
+
+from imodels.util.rule import Rule
+
+
+class SlipperRule(BaseEstimator, ClassifierMixin):
```
Comment (csinva) [inline imodels/rule_set/slipper.py:12]: Could you rename this to SlipperClassifier to be consistent with everything else (even if it's mainly just used by boosted_rules)?

File: imodels/rule_set/boosted_rules.py (hunk 1 above)
Comment (csinva) [reply imodels/rule_set/boosted_rules.py:22]: Agreed, it also makes it difficult to pass arguments (e.g. max_depth) to the estimator if it only takes a string. I think this argument should be preserved the way it is, so it doesn't break backward compatibility. In addition, if the argument to estimator is of type string, then check if it's "SLIPPER" and then set it the way you have used here.

File: imodels/rule_set/slipper.py (hunk 2 above)
Comment (csinva) [reply imodels/rule_set/slipper.py:250]: I think this is good!

**Comment by csinva** [comment]
Thanks for the really nice additions 😄 !

File: imodels/rule_set/boosted_rules.py (hunk 1 above)
Comment (sms1097) [reply imodels/rule_set/boosted_rules.py:22]: Do you think it's worth spending some time to wrap this closer to the procedure used in sklearn boosting algorithms? For example have the BoostedRulesClassifier inherit from [BaseWeightBoosting](https://github.com/scikit-learn/scikit-learn/blob/2beed55847ee70d363bdbfe14ee4401438fba057/sklearn/ensemble/_weight_boosting.py#L52) then extend the validate and fit functions and implement the _boost function for the BoostedRulesClassifier?

Comment (csinva) [reply imodels/rule_set/boosted_rules.py:22]: I think that might be overkill for what we have now (unless we implement many more boosting-type algorithms in the near future) - for now it's probably easiest to just have estimator be a flexible argument.

File: imodels/rule_set/boosted_rules.py [hunk 4]
```diff
@@ -133,13 +144,6 @@ def predict(self, X):
         X = check_array(X)
         return self.eval_weighted_rule_sum(X) > 0
 
-    # def predict(self, X):
-    #     """Predict outcome for X
-    #     """
-    #     check_is_fitted(self)
-    #     X = check_array(X)
-    #     return self.predict_proba(X).argmax(axis=1)
-
```
Comment (keyan3) [inline imodels/rule_set/boosted_rules.py:142]: Hey Sean, thanks for the contribution! Can you uncomment this instead of deleting it?

Comment (sms1097) [reply imodels/rule_set/boosted_rules.py:142]: Looks like the reason this was commented to update the predict function to use the `eval_weighted_rule_sum` function inherited from `RuleSet`. I can't uncomment because we would have conflicting definitions for the predict function. Do you want me to just leave it commented?

Comment (keyan3) [reply imodels/rule_set/boosted_rules.py:142]: I see — then deleting is ok. I think this was a reminder that we eventually want a `predict` that just takes the argmax of `predict_proba`, but we need to do some more work on `predict_proba` first. I'll add it as a todo item instead.

File: imodels/rule_set/boosted_rules.py (hunk 1 above)
Comment (sms1097) [reply imodels/rule_set/boosted_rules.py]: Yeah if there isn't any intention to immediately expand the flexibility of the boosting procedure then I would agree. I updated to accept the partial estimator and then check that object type once instantiated instead of the string

**Comment by csinva** [comment]
Looks great - I'll go ahead and merge it then - hope it's useful to a lot of folks :)

**Comment by csinva** [comment]
Hm the test for this model still seems to fail sometimes...

**Comment by sms1097** [comment]
Hmm that's weird, I'll investigate what's happening. I ran this a few times and didn't see any problems, but maybe something slipped back in that I missed

**Comment by sms1097** [comment]
Yeah so looking at this it fails to always score above 0.8 accuracy when n_estimators > 2. I didn't mess with the boosting procedure for slipper because I presumed it would be similar enough. Should I play with things to see if I can get a consistent score above 0.8?

**Comment by csinva** [comment]
Gotcha, so it passes when n_estimators = 2? I guess for the test we can just set it equal to 2 then.

**Comment by sms1097** [comment]
Actually I didn't run that enough, still fails occasionally for n_estimators greater than 2. I suppose that makes sense since the default rule could learn higher confidence than the sum of rules that span a smaller set of feature space. I can make a change that reflects that if what I just said makes sense.

**Comment by csinva** [comment]
Okay that sounds good for now! Also what is causing the stochasticity? The test is seeding np.random so it should all be deterministic (unless we need to seed something else as well)?

**Comment by sms1097** [comment]
The default rule randomly selects a random number of features randomly from the native random python module. If the default rule wins, which it does, it has a varying level of confidence on each run. On the first iteration Slipper will never pick a default rule so for n_estimators=1 we won't have the stochasticity problems we're seeing now.
